iceoryx_examples/complexdata/iox_publisher_vector.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/cxx/vector.hpp"
+
+#include "iceoryx_posh/popo/publisher.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/posix_wrapper/signal_handler.hpp"
+
+bool killswitch = false;
+constexpr char APP_NAME[] = "iox-cpp-publisher-vector";
+
+static void sigHandler(int f_sig [[gnu::unused]])
+{
+    // caught SIGINT or SIGTERM, now exit gracefully
+    killswitch = true;
+}
+
+int main()
+{
+    // register sigHandler
+    auto signalIntGuard = iox::posix::registerSignalHandler(iox::posix::Signal::INT, sigHandler);
+    auto signalTermGuard = iox::posix::registerSignalHandler(iox::posix::Signal::TERM, sigHandler);
+
+    // initialize runtime
+    iox::runtime::PoshRuntime::initRuntime(APP_NAME);
+
+    // initialize publisher
+    iox::popo::Publisher<iox::cxx::vector<double, 5>> publisher({"Radar", "FrontLeft", "Object"});
+
+    uint64_t ct = 0;
+    // run until interrupted by Ctrl-C
+    while (!killswitch)
+    {
+        publisher.loan()
+            .and_then([&](auto& sample) {
+                for (uint64_t i = 0U; i < sample->capacity(); ++i)
+                {
+                    bool success = sample->emplace_back(static_cast<double>(ct + i));
+                    if (!success)
+                    {
+                        std::cerr << "Failed to insert element." << std::endl;
```

Comment:
If this happens I think we should `break;` after the error message and send the stuff that was emplaced

Reply:
in this case I would omit the check since the loop doesn't exceed the capacity of the vector and this is unnecessary noise.

But we might think about having
```c++
cxx::vector<uint8_t, 42> foo;

uint8_t counter = 0;
foo.fill_to_capacity([&] {
    return counter++;
});

foo.for_each([](auto& element){
    std::cout << element << std::endl;
});
```

Reply:
This is one of the cases where we even know that the emplacement cannot fail anyway (barring a a wrong implementation of vector or loan which does not construct a sample with an empty vector, of course). So we should avoid unnecessary checks. 
Also take a look at https://en.cppreference.com/w/cpp/algorithm/iota for similar purposes (since we should even have random access iterators here it would apply). But it would have to operate on a non empty range, which we will not have initially here.

Reply:
@elBoberido we already have and `.for_each` at our hands.
```cpp
std::for_each(foo.begin(), foo.end(), [](auto e){std::cout << e << std::endl;});
```

Maybe we need a cxx version for it so that it is possible to do this for the whole range whenever it has `.end` and `.begin` as methods like:
```cpp
cxx::for_each(foo, [](auto e){std::cout << e << std::endl;});
```
This would be more like the C++ approach advertised in the STL.


The second problem I would solve like this:
```cpp
cxx::vector<uint8_t, 42> foo(42);
std::generate(foo.begin(), foo.end(), [counter=0]() mutable { return counter++; });
```
Maybe we also need a cxx version of that which would become more readable;
```cpp
cxx::vector<uint8_t, 42> foo(42);
cxx::generate(foo, [counter=0]() mutable { return counter++; });
```

But in the case that the elements are not default constructable we would have to pursue your approach since the containers have different method names to insert another element.

Reply:
removed the check

Reply:
you're example for generate won't work since `foo.begin() == foo.end()`

Reply:
@elBoberido, see the first line:
```cpp
cxx::vector<uint8_t, 42> foo(42); // this ctor creates 42 elements with the default ctor
```
This is a full vector and `foo.begin() != foo.end()`.

Reply:
@elfenpiff in this case you call the default ctor first, just to afterwards immediately replace the values. Not very efficient ;)